lib/go/test/flow_idtable_staking_test.go

```diff
@@ -2413,6 +2422,119 @@ func TestIDTableStaking(t *testing.T) {
 		)
 
 	})
+
+	t.Run("Should be able to enforce slot limits without ending staking", func(t *testing.T) {
+		//remove nodes from previous tests
+		currentNodes := [5]string{adminID, joshID, maxID, bastianID, accessID}
+
+		for i := 0; i < len(currentNodes); i++ {
+			tx := createTxWithTemplateAndAuthorizer(b, templates.GenerateRemoveNodeScript(env), idTableAddress)
+
+			_ = tx.AddArgument(CadenceString(currentNodes[i]))
+
+			signAndSubmit(
+				t, b, tx,
+				[]flow.Address{idTableAddress},
+				[]crypto.Signer{IDTableSigner},
+				false,
+			)
+		}
+
+		//update the allow list
+		nodeCount := 15
+		allowList := make([]cadence.Value, nodeCount)
+		for i := 0; i < nodeCount; i++ {
+			allowList[i] = CadenceString(fmt.Sprintf("%064d", i))
+		}
+
+		tx := createTxWithTemplateAndAuthorizer(b, templates.GenerateSetApprovedNodesScript(env), idTableAddress)
+
+		err := tx.AddArgument(cadence.NewArray(allowList))
+		require.NoError(t, err)
+
+		signAndSubmit(
+			t, b, tx,
+			[]flow.Address{idTableAddress},
+			[]crypto.Signer{IDTableSigner},
+			false,
+		)
+
+		//prepare the arrays for bulk node registration
+		authorizers := make([]flow.Address, nodeCount)
+		signers := make([]crypto.Signer, nodeCount)
+		stakingKeys := make([]string, nodeCount)
+		networkingKeys := make([]string, nodeCount)
+		ids := make([]string, nodeCount)
+
+		//register nodeCount nodes for the first test
+		for i := 0; i < nodeCount; i++ {
+			authorizers[i] = jessAddress		//array of authorizers - flow.address
+			signers[i] = jessSigner		//array of signers - crypto.signer
+			_, stakingKeys[i], _, networkingKeys[i] = generateKeysForNodeRegistration(t)
+			ids[i] = fmt.Sprintf("%064d", i)	//array of ids - string
+		}
+
+		registerNodesForStaking(t, b, env,
+			authorizers,
+			signers,
+			stakingKeys,
+			networkingKeys,
+			ids)
+
+		//check resulting array length
+		result := executeScriptAndCheck(t, b, templates.GenerateReturnProposedTableScript(env), nil)
+		idArray := result.(cadence.Array).Values
+		assert.Len(t, idArray, nodeCount)
+
+		//remove unapproved nodes to match final state
```

Comment:
What does "match final state" mean here?

Reply:
We envisioned that slot selection would run after removeUnapprovedNodes. To prepare for the test I create a series of new nodes. It is necessary to run "removeUnapprovedNodes" to get the system into the final state that is representative of what it would be before slot selection. Under the hood it's setting all of the initialWeight values, but I preferred to do that with the built-in function rather than forcing it.